crates/metal-llvm/src/stdlib_priv.rs

```diff
@@ -0,0 +1,82 @@
+use std::{ffi::CString, ptr};
+
+use llvm_sys::{
+    core::LLVMCreateMemoryBufferWithMemoryRangeCopy, ir_reader::LLVMParseIRInContext, prelude::*,
+};
+
+// NOTE: private modules do NOT have submodules.
+/// private modules stand at the core of the stdlib.
+/// They allow us to define metal functionality in the stdlib
+/// for compile time.
+pub trait PrivateSTDLibModule {
+    // private modules usually start with an underscore
+    // But this name does not include that underscore.
+    /// Private STDLib module name.
+    /// i.e. `io`, `cmplib`, etc.
+    fn libname() -> &'static str;
+
+    fn function(
+        ident: metal_ast::Ident,
+        ctx: LLVMContextRef,
+        module: LLVMModuleRef,
+        arguments: Vec<String>,
+    ) -> Option<LLVMValueRef>;
+}
+
+pub struct ExternalLib;
+
+impl PrivateSTDLibModule for ExternalLib {
+    fn libname() -> &'static str {
+        "pext"
+    }
+
+    #[allow(clippy::not_unsafe_ptr_arg_deref)]
+    fn function(
+        ident: metal_ast::Ident,
+        ctx: LLVMContextRef,
+        module: LLVMModuleRef,
+        arguments: Vec<String>,
+    ) -> Option<LLVMValueRef> {
+        match ident.inner {
+            "extern" => {
+                if arguments.len() != 3 {
+                    panic!("extern does not accept more or less than 3 arguments")
+                }
+                let name = &arguments[0];
+                let ret_type = &arguments[1];
+                let args = &arguments[2];
+
+                let ir = format!(
+                    "
+                    define {} @{}({}) unnamed_addr
+                ",
+                    ret_type, name, args
+                );
+
+                unsafe {
+                    let ir_buffer = LLVMCreateMemoryBufferWithMemoryRangeCopy(
+                        ir.as_ptr() as *const i8,
+                        ir.len(),
+                        "ir_buffer".as_ptr() as *const i8,
+                    );
+
+                    let mut error = ptr::null_mut();
+
+                    let suc =
+                        LLVMParseIRInContext(ctx, ir_buffer, [module].as_mut_ptr(), &mut error);
```

Comment:
I'm not sure the third parameter should be a slice. [inkwell source](https://thedan64.github.io/inkwell/src/inkwell/context.rs.html#94) (called from [here](https://thedan64.github.io/inkwell/inkwell/context/struct.Context.html#method.create_module_from_ir)) suggests it should be a nullptr. I couldn't decipher the meaning myself, but [here](https://llvm.org/doxygen/group__LLVMCCoreIRReader.html#ga319e4562ffb47ec6eba2eb70ffdbaa37) is the LLVM source for this function. My intuition tells that LLVM expects a mut ptr to pre-allocated memory the size of a module, which seems to be what calling `null_mut` does [through `MaybeUninit`](https://doc.rust-lang.org/std/mem/union.MaybeUninit.html#method.zeroed)

Reply:
What has drawn you into that conclusion?

Reply:
> What has drawn you into that conclusion?

Actually nevermind, it seems you're actually right. Looking at LLVM Code (https://github.com/hdoc/llvm-project/blob/release/15.x/llvm/lib/IRReader/IRReader.cpp#L116-L117) it seems to set the pointer which most likely means it's making a new module.

Reply:
It should be noted that my docs are for LLVM 20 and yours are for LLVM 15 (and we use LLVM 18), however, given that this function went unchanged for five major releases, it should be safe to assume LLVM 18 does the same.